django/db/models/base.py

```diff
@@ -454,10 +465,12 @@ def __init__(self, *args, **kwargs):
 
     @classmethod
     def from_db(cls, db, field_names, values):
-        if cls._deferred:
-            new = cls(**dict(zip(field_names, values)))
-        else:
-            new = cls(*values)
+        # TODO - make QuerySet iterator fill the missing fields.
```

Comment:
I see. Not sure if this is what you had in mind, but here's what I tried:
```diff
diff --git a/django/db/models/base.py b/django/db/models/base.py
index 1912337..2e191ec 100644
--- a/django/db/models/base.py
+++ b/django/db/models/base.py
@@ -465,11 +465,6 @@ class Model(six.with_metaclass(ModelBase)):
 
     @classmethod
     def from_db(cls, db, field_names, values):
-        # TODO - make QuerySet iterator fill the missing fields.
-        if len(values) != len(cls._meta.concrete_fields):
-            values = list(values)
-            values.reverse()
-            values = [values.pop() if f.attname in field_names else DEFERRED for f in cls._meta.concrete_fields]
         new = cls(*values)
         new._state.adding = False
         new._state.db = db
diff --git a/django/db/models/query.py b/django/db/models/query.py
index 03d12cf..1d63803 100644
--- a/django/db/models/query.py
+++ b/django/db/models/query.py
@@ -45,6 +45,8 @@ class ModelIterable(BaseIterable):
     """
 
     def __iter__(self):
+        from django.db.models.base import DEFERRED
+
         queryset = self.queryset
         db = queryset.db
         compiler = queryset.query.get_compiler(using=db)
@@ -61,8 +63,17 @@ class ModelIterable(BaseIterable):
         init_list = [f[0].target.attname
                      for f in select[model_fields_start:model_fields_end]]
         related_populators = get_related_populators(klass_info, select, db)
+        fields_are_deferred = len(init_list) != len(model_cls._meta.concrete_fields)
         for row in compiler.results_iter(results):
-            obj = model_cls.from_db(db, init_list, row[model_fields_start:model_fields_end])
+            values = row[model_fields_start:model_fields_end]
+            if fields_are_deferred:
+                values = list(values)
+                values.reverse()
+                values = [
+                    values.pop() if f.attname in init_list else DEFERRED
+                    for f in model_cls._meta.concrete_fields
+                ]
+            obj = model_cls.from_db(db, init_list, values)
             if related_populators:
                 for rel_populator in related_populators:
                     rel_populator.populate(row, obj)
```
This yields two test failures in the `defer` test app for tests involving `select_related()` (didn't check any other test apps). There are a couple other `from_db()` calls in `query.py` so I guess those would need to be updated to have similar logic to supply all fields. I guess it makes sense to keep that logic in one place rather than duplicate it, but let me know if you had something else in mind. We could always create a separate ticket for investigating this more in depth.